src/FsAutoComplete.Core/State.fs

```diff
@@ -38,6 +41,15 @@ type State =
     x.Files.[file] <- fileState
     x.FileCheckOptions.[file] <- opts
 
+  member x.AddCancellationToken(file : SourceFilePath, token: CancellationTokenSource) =
+    x.CancellationTokens.AddOrUpdate(file, [token], fun _ lst -> token::lst)
+    |> ignore
+
+  member x.GetCancellationTokens(file : SourceFilePath) =
+    let lst = x.CancellationTokens.GetOrAdd(file, fun _ -> [])
+    x.CancellationTokens.TryRemove(file) |> ignore
+    lst
```

Comment:
I'd suggest

```
member x.GetCancellationTokens(file : SourceFilePath) =
   match x.CancellationTokens.TryRemove(file) with
   | true, lst -> lst
   | _ -> []
```

Which should remove the race.

Reply:
Oh damn, obviously. Not sure how I've missed that/